framework/include/systems/NonlinearSystemBase.h

```diff
@@ -227,6 +227,45 @@ class NonlinearSystemBase : public SolverSystem, public PerfGraphInterface
    */
   std::shared_ptr<Split> getSplit(const std::string & name);
 
+  /**
+   * We offer the option to check convergence against the pre-SMO residual. This method handles the
+   * logic as to whether we should perform such residual evaluation.
+   *
+   * @return A boolean indicating whether we should evaluate the pre-SMO residual.
+   */
+  bool shouldEvaluatePreSMOResidual() const;
+
+  /**
+   * Returns a writable reference to a boolean controlling whether to evaluate the pre-SMO residual
+   * and use it in the subsequent relative convergence checks.
+   *
+   * If set to true, an _additional_ residual evaluation is performed before any
+   * solution-modifying object is executed, and before the initial (0-th nonlinear iteration)
+   * residual evaluation. Such residual is referred to as the pre-SMO residual. If the pre-SMO
+   * residual is evaluted, it is used in the subsequent relative convergence checks.
+   *
+   * If set to false, no residual evaluation takes place before the initial residual evaluation, and
+   * the initial residual is used in the subsequent relative convergence checks. This mode is
+   * recommended for performance-critical code as it avoids the additional pre-SMO residual
+   * evaluation.
+   */
+  bool & usePreSMOResidual() { return _use_pre_smo_residual; }
```

Comment:
In principle we don't need both methods, because `shouldEvaluatePreSMOResidual()` and `usePreSMOResidual()` should always return the same value. However, we need both for now to stay consistent with the legacy behavior (where an additional residual evaluation always takes place).